### PR TITLE
recusively query window tree

### PR DESCRIPTION
**Pull request by jstkdng**
After seebye release version 18.1.9 , he switched from python-xlib to xres and xshm. This change caused the ueberzug window to not appear in certain conditions where the ueberzug process was deeply nested. as in
`suckless tabbed -> tmux -> ueberzug` or `i3 tabs -> tmux -> ueberzug`, this is caused because the window tree was only being used to get the first level of windows and not any child windows.

This PR fixes this issue.

### CODE REVIEW

File: ueberzug/X/display.c [hunk 1]
```diff
@@ -92,25 +93,50 @@ has_property(DisplayObject *self, Window window, Atom property) {
...
+get_child_window_ids_helper(Display *display, Window window,
+        Window **total_children, unsigned int *total_num_children) {
+    Window _, *children;
+    unsigned int num_children;
+
+    XQueryTree(display, window, &_, &_, &children, &num_children);
+    if (!children) return;
+
+    if (!*total_children) {
+        *total_children = calloc(num_children, sizeof(Window));
```
Comment (N-R-K) [inline ueberzug/X/display.c:106]: No need for this branch, if `ptr` is NULL, `realloc` acts as `malloc`.

File: ueberzug/X/display.c [hunk 2]
```diff
@@ -92,25 +93,50 @@ has_property(DisplayObject *self, Window window, Atom property) {
...
+    } else {
+        Window *tmp = reallocarray(*total_children, num_children + *total_num_children, sizeof(Window));
```
Comment (N-R-K) [inline ueberzug/X/display.c:108]: If `realloc` fails, won't this go out of bound later down? Also `reallocarray` is a non-standard BSD extension. Don't know what the policy is on the ueuberzug codebase towards using such extensions.

File: ueberzug/X/display.c [hunk 3]
```diff
@@ -92,25 +93,50 @@ has_property(DisplayObject *self, Window window, Atom property) {
...
             &parent)) {
         Py_RETURN_ERROR;
     }
 
-    if (!XQueryTree(
-            self->info_display, parent,
-            &_, &_, &children, &children_count)) {
-        raise(OSError, "failed to query child windows of %lu", parent);
-    }
+    get_child_window_ids_helper(self->info_display, parent, &children, &children_count);
```
Comment (N-R-K) [inline ueberzug/X/display.c:139]: The `children` array is being passed to XFree later down, which seems like a wrong thing to do because it wasn't allocated via XLib. (It will *happen* to work because XFree calls `free` internally, but I don't think we should depend on XLib's internal behaviors).

**Review by N-R-K — CHANGES_REQUESTED** [review]
It would be nice if the commit msg described what *effect* the patch has (i.e what difference will it make to the end-user).

File: ueberzug/X/display.c (hunk 2 above)
Comment (jstkdng) [reply ueberzug/X/display.c:108]: >Also reallocarray is a non-standard BSD extension.

is it?

```
reallocarray():
           Since glibc 2.29:
               _DEFAULT_SOURCE
           Glibc 2.28 and earlier:
               _GNU_SOURCE
```

File: ueberzug/X/display.c (hunk 3 above)
Comment (jstkdng) [reply ueberzug/X/display.c:139]: fair enough

File: ueberzug/X/display.c (hunk 2 above)
Comment (N-R-K) [reply ueberzug/X/display.c:108]: Those are just feature test macros. See the CONFORMING TO section:

```
CONFORMING TO
       malloc(), free(), calloc(), realloc(): POSIX.1-2001, POSIX.1-2008, C89, C99.

       reallocarray() is a nonstandard extension that first appeared in OpenBSD 5.6 and FreeBSD 11.0.
```

Comment (jstkdng) [reply ueberzug/X/display.c:108]: I see, currently there is no explicit C standard enforced in the code (gcc's default perhaps?).

**Comment by jstkdng** [comment]
Now that's what I call unreadable.

>rc |= get_child_window_ids_helper

why bitwise operators?

**Comment by eylles** [comment]
sorry work has been kicking my butt so i haven't had time to look at this, will try to do this week tho i suppose this is the evolution of the patches to make ueberzug work with tabbed, tmux, urxvt, mate-terminal, etc... right?

**Comment by jstkdng** [comment]
tabbed and tmux are confirmed to work. Don't know about the other ones, they should though.

**Comment by eylles** [comment]
will try to give this a test run today, see if i can find some bug or edge case

also, there was a patch for xutil.py to make it work with urxvt mate-terminal and so on that run as a single process but may have many windows

https://github.com/pld-linux/ueberzug/blob/master/single_process_terminals.patch

**Comment by jstkdng** [comment]
That patch is really old, but it is equivalent to this change. https://github.com/gokberkgunes/ueberzug-tabbed/commit/6aafc5e2d12d23996e4c20f7f45ef12ce784c087

**Comment by eylles** [comment]
yeh, honestly i have not tested daemon mode with urxvt nor mate terminal, will try to see what happens with urxvt and report, rn not at home but will begin tests once i'm in, i want to merge this between today and tomorrow since i got the time thanks to a sick leave from work.
